scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+# Here I have declared them as taking void* instead of PyArrayDescr
+# and object. In this file, only NULL is passed to these parameters.
+cdef extern from *:
+    cnp.ndarray PyArray_CheckFromAny(object, void*, int, int, int, void*)
+    cnp.ndarray PyArray_FromArray(cnp.ndarray, void*, int)
+
+from scipy.linalg cimport blas_pointers
+from scipy.linalg cimport lapack_pointers
+
+import numpy as np
+
+#------------------------------------------------------------------------------
+# These are a set of fused type wrappers around the BLAS and LAPACK calls used. 
+#------------------------------------------------------------------------------
+
+ctypedef float complex float_complex
+ctypedef double complex double_complex
+ctypedef fused blas_t:
+    float
+    double
+    float_complex
+    double_complex
+
+cdef inline blas_t* index2(blas_t* a, int* as, int i, int j) nogil:
+    return a + i*as[0] + j*as[1]
+
+cdef inline blas_t* index1(blas_t* a, int* as, int i) nogil:
+    return a + i*as[0]
+
+cdef inline blas_t* row(blas_t* a, int* as, int i) nogil:
+    return a + i*as[0]
+
+cdef inline blas_t* col(blas_t* a, int* as, int j) nogil:
+    return a + j*as[1]
+    
+cdef inline void copy(int n, blas_t* x, int incx, blas_t* y, int incy) nogil:
+    if blas_t is float:
+        blas_pointers.scopy(&n, x, &incx, y, &incy) 
+    elif blas_t is double:
+        blas_pointers.dcopy(&n, x, &incx, y, &incy)
+    elif blas_t is float_complex:
+        blas_pointers.ccopy(&n, x, &incx, y, &incy)
+    else:
+        blas_pointers.zcopy(&n, x, &incx, y, &incy)
+
+cdef inline void swap(int n, blas_t* x, int incx, blas_t* y, int incy) nogil:
+    if blas_t is float:
+        blas_pointers.sswap(&n, x, &incx, y, &incy) 
+    elif blas_t is double:
+        blas_pointers.dswap(&n, x, &incx, y, &incy)
+    elif blas_t is float_complex:
+        blas_pointers.cswap(&n, x, &incx, y, &incy)
+    else:
+        blas_pointers.zswap(&n, x, &incx, y, &incy)
+
+cdef inline void scal(int n, blas_t a, blas_t* x, int incx) nogil:
+    if blas_t is float:
+        blas_pointers.sscal(&n, &a, x, &incx)
+    elif blas_t is double:
+        blas_pointers.dscal(&n, &a, x, &incx)
+    elif blas_t is float_complex:
+        blas_pointers.cscal(&n, &a, x, &incx)
+    else:
+        blas_pointers.zscal(&n, &a, x, &incx)
+
+cdef inline void axpy(int n, blas_t a, blas_t* x, int incx,
+                      blas_t* y, int incy) nogil:
+    if blas_t is float:
+        blas_pointers.saxpy(&n, &a, x, &incx, y, &incy)
+    elif blas_t is double:
+        blas_pointers.daxpy(&n, &a, x, &incx, y, &incy)
+    elif blas_t is float_complex:
+        blas_pointers.caxpy(&n, &a, x, &incx, y, &incy)
+    else:
+        blas_pointers.zaxpy(&n, &a, x, &incx, y, &incy)
+
+cdef inline blas_t nrm2(int n, blas_t* x, int incx) nogil:
+    # since dnrm2 etc are functions, we can't at present call them...
+    # this is mostly cribbed from netlib's dnrm2
```

Comment:
It's only really possible to call a fortran subroutine from C.  dnrm2 is a function and at the time #4021 hadn't wrapped any of the functions.  This will become something like the rest of the fused type wrappers now that that dnrm2 is wrapped.